src/components/DataActionButton/index.tsx

```diff
@@ -64,6 +86,182 @@ function DataActionButton(props: Props) {
+  for (const signed_url of signed_urls) {
+    const { volume, path, presigned_url } = signed_url;
+    const gdsUrl = constructGDSUrl({ volume_name: volume, path: path });
+    if (gdsUrl === fileGdsUrl) {
+      filePresignUrl = presigned_url;
+    } else if (gdsUrl === idxFileGdsUrl) {
+      idxFilePresignUrl = presigned_url;
+    }
+  }
+
+  const idx = encodeURIComponent(idxFilePresignUrl);
+  const enf = encodeURIComponent(filePresignUrl);
+  const name = pathOrKey.split('/').pop() ?? pathOrKey;
+  return `http://localhost:60151/load?index=${idx}&file=${enf}&name=${name}`;
+};
+
+const constructS3LocalIgvUrl = async (props: {
+  id: number;
+  bucketOrVolume: string;
+  pathOrKey: string;
+}) => {
+  const { id, bucketOrVolume, pathOrKey } = props;
+
+  const objStatus = await getS3Status(id);
+  if (objStatus != S3StatusData.AVAILABLE) {
+    throw new Error(`Invalid object storage class! (${S3StatusData[objStatus]})`);
+  }
+
+  const name = pathOrKey.split('/').pop() ?? pathOrKey;
+  const file = `s3://${bucketOrVolume + '/' + pathOrKey}`;
+
+  return `http://localhost:60151/load?file=${encodeURIComponent(file)}&name=${name}`;
+};
+
+type OpenIGVDesktopType = DataActionButtonProps & { handleIsOpen: (val: boolean) => void };
+function OpenIGVDesktop(props: OpenIGVDesktopType) {
+  const toast = useToastContext();
+
+  const { id, bucketOrVolume, pathOrKey, type, handleIsOpen } = props;
+
+  const gdsLocalIgvUrl = useQuery(
+    ['gds-local-igv', bucketOrVolume, pathOrKey],
+    async () =>
+      await constructGDSLocalIgvUrl({ bucketOrVolume: bucketOrVolume, pathOrKey: pathOrKey }),
+    { enabled: type == 'gds' }
+  );
+
+  const s3LocalIgvUrl = useQuery(
+    ['s3-local-igv', bucketOrVolume, pathOrKey],
+    async () =>
+      await constructS3LocalIgvUrl({
+        id: id,
+        bucketOrVolume: bucketOrVolume,
+        pathOrKey: pathOrKey,
+      }),
+    { enabled: type == 's3' }
+  );
+
+  if (gdsLocalIgvUrl.isLoading || s3LocalIgvUrl.isLoading) {
+    return (
+      <Dialog
+        header='Opening in IGV Desktop'
+        visible={true}
+        style={{ width: '50vw' }}
+        draggable={false}
+        resizable={false}
+        onHide={() => handleIsOpen(false)}>
+        <CircularLoaderWithText />
+      </Dialog>
+    );
+  }
+
+  if (s3LocalIgvUrl.isError && s3LocalIgvUrl.error) {
+    toast?.show({
+      severity: 'error',
+      summary: 'Error on locating object URL.',
+      detail: `${s3LocalIgvUrl.error}`,
+      sticky: true,
+    });
+  }
+  if (gdsLocalIgvUrl.isError && gdsLocalIgvUrl.error) {
+    toast?.show({
+      severity: 'error',
+      summary: 'Error on locating object URL.',
+      detail: `${gdsLocalIgvUrl.error}`,
+      sticky: true,
+    });
+  }
+
+  const xhr = new XMLHttpRequest();
+
+  let localIgvUrl = '';
+
+  localIgvUrl = type == 'gds' && gdsLocalIgvUrl.data ? gdsLocalIgvUrl.data : '';
+  localIgvUrl = type == 's3' && s3LocalIgvUrl.data ? s3LocalIgvUrl.data : '';
```

Comment:
This second assignment would override/mutate `localIgvUrl`. We might just want to simply fallback the classic if/else guard here, for example.